java/lang/security/audit/xxe/documentbuilderfactory-disallow-doctype-decl-missing.yaml

```diff
@@ -0,0 +1,47 @@
+rules:
+  - id: documentbuilderfactory-disallow-doctype-decl-missing
+    severity: ERROR
+    metadata:
+      cwe: "CWE-611: Improper Restriction of XML External Entity Reference"
+      owasp: "A4: XML External Entities (XXE)"
+      asvs:
+        section: V5 Validation, Sanitization and Encoding
+        control_id: 5.5.2 Insecue XML Deserialization
+        control_url: https://github.com/OWASP/ASVS/blob/master/4.0/en/0x13-V5-Validation-Sanitization-Encoding.md#v55-deserialization-prevention
+        version: "4"
+      references:
+        - https://blog.sonarsource.com/secure-xml-processor
+        - https://xerces.apache.org/xerces2-j/features.html
+      category: security
+      technology:
+        - java
+    message: >-
+      DOCTYPE declaratons are enabled for this DocumentBuilderFactory.
+      This is vulnerable to XML external entity attacks. Disable this by setting the feature "http://apache.org/xml/features/disallow-doctype-decl" to true.
+      Alternatively, allow DOCTYPE declarations and only prohibit external entities declarations.
+      This can be done by setting the features "http://xml.org/sax/features/external-general-entities" and "http://xml.org/sax/features/external-parameter-entities" to false.
+    patterns:
+      - pattern: DocumentBuilderFactory $DBF = DocumentBuilderFactory.newInstance();
+      - pattern-not-inside: |
+          $RETURNTYPE $METHOD(...){
+            ...
+            $DBF.setFeature("http://apache.org/xml/features/disallow-doctype-decl", ...);
+            ...
+          }
+      - pattern-not-inside: |
+          $RETURNTYPE $METHOD(...){
+            ...
+            $DBF.setFeature("http://xml.org/sax/features/external-general-entities", ...);
+            ...
+          }
+      - pattern-not-inside: |
+          $RETURNTYPE $METHOD(...){
+            ...
+            $DBF.setFeature("http://xml.org/sax/features/external-parameter-entities", ...);
+            ...
+          }
+    fix: |
+        DocumentBuilderFactory $DBF = DocumentBuilderFactory.newInstance();
+        $DBF.setFeature("http://apache.org/xml/features/disallow-doctype-decl", true);
+    languages:
+      - java
```

Comment:
might be worth adding `xml` technology tag

Reply:
xml can be added to technology but not to languages.